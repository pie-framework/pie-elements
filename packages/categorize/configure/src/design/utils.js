export const getPluginProps = (props = {}, baseInputConfiguration = {}) => ({
  ...baseInputConfiguration,
  ...props,
});
