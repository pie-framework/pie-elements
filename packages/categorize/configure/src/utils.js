export const generateValidationMessage = (config) => {
  let { minChoices, maxChoices, maxCategories } = config || {};
  minChoices = minChoices || 1;

  const validationMessage =
    '\nThe choices content should ' +
    'not be empty and should be unique.\nThere should be at least 1 ' +
    (maxCategories ? `and at most ${maxCategories} ` : '') +
    'category' +
    (maxCategories ? 's' : '') +
    ' defined.' +
    (minChoices ? `\nThere should be at least ${minChoices} choices defined.` : '') +
    (maxChoices ? `\nNo more than ${maxChoices} choices should be defined.` : '') +
    '\nAt least one token should be assigned to at least one category.';

  return 'Validation requirements:' + validationMessage;
};

// used in controller too, for consistency modify it there too
export const multiplePlacements = { enabled: 'Yes', disabled: 'No', perChoice: 'Set Per Choice' };
