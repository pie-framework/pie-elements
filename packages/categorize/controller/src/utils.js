// used in configure too, for consistency modify it there too
export const multiplePlacements = { enabled: 'Yes', disabled: 'No', perChoice: 'Set Per Choice' };

// used to validate the config
export const isCorrectResponseDuplicated = (choices,alternate) => {
    const stringChoices =  JSON.stringify(choices.sort());
    const stringAlternate = alternate.filter(alternate => alternate.length > 0).map((alternate) => JSON.stringify(alternate.sort()));
    const foundIndexDuplicate = stringAlternate.findIndex(alternate => alternate.length &&  alternate === stringChoices);
    return foundIndexDuplicate;
};

export const isAlternateDuplicated = (alternate) => {
    const elementSet = new Set();
    const stringAlternate = alternate.filter(alternate=>alternate.length > 0).map((alternate) => JSON.stringify(alternate.sort()));
    for (let i = 0; i < stringAlternate.length; i++) {
        if (elementSet.has(stringAlternate[i]) && stringAlternate[i]) {
            return i;
        }
        elementSet.add(stringAlternate[i]);
    }

    return -1;
};
