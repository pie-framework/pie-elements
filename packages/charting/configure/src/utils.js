// this file is almost duplicated from graphing/configure/src/utils.js; the same logic must be used in charting
import isEqual from 'lodash/isEqual';

const VALID_GRID_VALUES = [
  0.01, 0.02, 0.04, 0.05, 0.0625, 0.1, 0.125, 0.2, 0.25, 0.5, 1, 2, 3, 4, 5, 8, 10, 12, 15, 20, 40, 50, 64, 100, 500,
  1000,
];

// same as VALID_GRID_VALUES but next values are excluded: 0.04, 0.0625, 0.125, 3, 4, 8, 12, 15, 40, 64
const PREFERRED_VALID_GRID_VALUES = [0.01, 0.02, 0.05, 0.1, 0.2, 0.25, 0.5, 1, 2, 5, 10, 20, 50, 100, 500, 1000];

export const VALID_LABEL_VALUES = {
  0.01: [0, 0.01, 0.02, 0.04, 0.05, 0.1],
  0.02: [0, 0.02, 0.04, 0.1],
  0.04: [0, 0.04, 0.08, 0.16, 0.2],
  0.05: [0, 0.05, 0.1, 0.2, 0.25],
  0.0625: [0, 0.0625, 0.125, 0.25, 0.5],
  0.1: [0, 0.1, 0.2, 0.4, 0.5, 1],
  0.125: [0, 0.125, 0.25, 0.5, 1],
  0.2: [0, 0.2, 0.5, 0.1],
  0.25: [0, 0.25, 0.5, 1, 2],
  0.5: [0, 0.5, 1, 2],
  1: [0, 1, 2, 4, 5, 10],
  2: [0, 2, 4, 8, 10],
  3: [0, 3, 6, 12, 15],
  4: [0, 4, 8, 16, 20],
  5: [0, 5, 10, 20, 25],
  8: [0, 8, 16, 32, 40, 64],
  10: [0, 10, 20, 40, 50, 100],
  12: [0, 12, 24],
  15: [0, 15, 30, 60],
  20: [0, 20, 40, 80, 100],
  40: [0, 40, 80, 160, 200],
  50: [0, 50, 100, 200, 250],
  64: [0, 64, 128],
  100: [0, 100, 200, 400, 500],
  500: [0, 500, 1000, 2500],
  1000: [0, 1000, 2000, 4000, 5000],
};

export const getGridValues = (axis, size, prefferedValues = false) => {
  const minValue = (10 * (axis.max - axis.min)) / size;
  const maxValue = minValue * 10;
  const values = prefferedValues ? PREFERRED_VALID_GRID_VALUES : VALID_GRID_VALUES;

  return values.filter((value) => value >= minValue && value <= maxValue);
};

export const getLabelValues = (value) => VALID_LABEL_VALUES[value] || [];

export const applyConstraints = (axis, size, oldGridValues, oldLabelValues) => {
  const gridValues = getGridValues(axis, size);
  let labelValues = getLabelValues(axis.step || 1);

  if (!isEqual(oldGridValues, gridValues) && !gridValues.includes(axis.step)) {
    const preferredValues = getGridValues(axis, size, true);
    const lowestValue = preferredValues[0] || 1;

    axis.step = lowestValue;
    labelValues = getLabelValues(lowestValue);

    if (!labelValues.includes(axis.labelStep)) {
      axis.labelStep = lowestValue;
    }

    return { gridValues, labelValues };
  }

  if (!isEqual(oldLabelValues, labelValues) && !labelValues.includes(axis.labelStep)) {
    axis.labelStep = axis.step;
  }

  return { gridValues, labelValues };
};
