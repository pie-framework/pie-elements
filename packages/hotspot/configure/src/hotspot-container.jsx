import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles/index';
import Help from '@material-ui/icons/Help';

import Drawable from './hotspot-drawable';
import Button from './button';

const isImage = (file) => {
  const imageType = /image.*/;
  return file.type.match(imageType);
};

class Container extends Component {
  constructor(props) {
    super(props);
    this.state = {
      maxImageWidth: 0,
      maxImageHeight: 0,
      dragEnabled: true
    }
  }

  componentDidMount() {
    const positionInfo = this.imageSection.getBoundingClientRect();
    const { height, width } = positionInfo;
    this.setState({
      maxImageWidth: width - 20,
      maxImageHeight: height - 20
    })
  }

  handleFileRead = (file) => {
    const { onImageUpload } = this.props;
    const reader = new FileReader();
    reader.onloadend = () => {
      onImageUpload(reader.result);
    };
    reader.readAsDataURL(file)
  };

  handleUploadImage = (e) => {
    e.preventDefault();
    const file = e.target.files[0];
    this.handleFileRead(file);
  };

  enableDropzone = () => this.setState({ dropzoneActive: true });
  disableDropzone = () => this.setState({ dropzoneActive: false });

  handleOnDrop = (e) => {
    e.preventDefault();
    const { items, files } = e.dataTransfer;

    if (items && items[0].kind === 'file') {
      const file = items[0].getAsFile();
      if (isImage(file)) {
        this.handleFileRead(file);
      }
    } else if (isImage(files[0])) {
      this.handleFileRead(files[0]);
    }
    this.disableDropzone();
  };

  handleOnDragOver = (e) => {
    e.preventDefault();
    const { dropzoneActive } = this.state;

    if (!dropzoneActive) {
      this.enableDropzone();
    }
  };

  handleOnDragExit = (e) => {
    e.preventDefault();
    this.disableDropzone();
  };

  handleUndo = () => {
    const { onUpdateShapes, shapes } = this.props;

    if (shapes.length) {
      const newShapes = shapes.slice(0, shapes.length - 1);
      onUpdateShapes(newShapes);
    }
  };

  handleClearAll = () => {
    const { onUpdateShapes } = this.props;
    onUpdateShapes([]);
  };

  handleEnableDrag = () => this.setState({ dragEnabled: true });
  handleDisableDrag = () => this.setState({ dragEnabled: false });
  handleInputClick = () => this.input.click();

  toggleTooltip = () => {
    const { showTooltip } = this.state;
    this.setState({ showTooltip: !showTooltip })
  };

  renderUploadControl(label) {
    const { classes } = this.props;

    return (
      <div>
        <Button
          label={label}
          onClick={this.handleInputClick}
        />
        <input
          accept="image/*"
          className={classes.input}
          onChange={this.handleUploadImage}
          ref={ref => { this.input = ref; }}
          type="file"
        />
      </div>
    )
  }

  render() {
    const {
      classes,
      hotspotColor,
      imageUrl,
      multipleCorrect,
      onUpdateImageDimension,
      onUpdateShapes,
      outlineColor,
      shapes
    } = this.props;
    const {
      dropzoneActive,
      dragEnabled,
      maxImageHeight,
      maxImageWidth,
      showTooltip
    } = this.state;

    return (
      <div className={classes.base} >
        <div className={`${classes.box} ${dropzoneActive ? classes.boxActive : ''}`}
             {...dragEnabled ? {
               onDragExit: this.handleOnDragExit,
               onDragLeave: this.handleOnDragExit,
               onDragOver: this.handleOnDragOver ,
               onDrop: this.handleOnDrop
             } : {}}
        >
          <div className={classes.toolbar}>
            <Button
              disabled={!shapes.length}
              onClick={this.handleUndo}
              label="Undo"
            />
            <Button
              disabled={!shapes.length}
              onClick={this.handleClearAll}
              label="Clear all"
            />
          </div>

          <div ref={ref => { this.imageSection = ref; }} className={classes.drawableHeight}>
            {imageUrl ? (
              <Drawable
                disableDrag={this.handleDisableDrag}
                enableDrag={this.handleEnableDrag}
                imageUrl={imageUrl}
                hotspotColor={hotspotColor}
                maxImageHeight={maxImageHeight}
                maxImageWidth={maxImageWidth}
                multipleCorrect={multipleCorrect}
                onUpdateImageDimension={onUpdateImageDimension}
                onUpdateShapes={onUpdateShapes}
                outlineColor={outlineColor}
                shapes={shapes}
              />
            ) : (
              <div className={`${classes.drawableHeight} ${classes.centered}`}>
                <label>Drag and drop or upload image from computer</label>
                <br />
                {this.renderUploadControl('Upload Image')}
              </div>
            )}
          </div>

          {imageUrl ? (
            <div className={classes.tooltip}>
              {showTooltip ? (
                <div className={classes.tooltipContent}>
                  <label>
                    Click, move mouse and click again to create a hotspot. Click the hotspot to mark correct. Click again to unmark.
                  </label>
                  <div className={classes.tooltipArrow} />
                </div>
              ) : null}
              <Help
                className={classes.icon}
                onMouseOut={this.toggleTooltip}
                onMouseOver={this.toggleTooltip}
              />
            </div>
          ) : null}
        </div>

        {imageUrl ? (
          <div className={classes.replace}>
            {this.renderUploadControl('Replace Image')}
          </div>
        ) : null}
      </div>

    );
  }
}

const styles = theme => ({
  base: {
    marginTop: theme.spacing.unit * 2
  },
  box: {
    border: '1px solid #E0E1E6',
    borderRadius: '5px'
  },
  boxActive: {
    border: '1px solid #0032C2'
  },
  centered: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center'
  },
  drawableHeight: {
    minHeight: 350
  },
  icon: {
    '&:hover': {
      color: '#333131',
      cursor: 'help'
    },
    color: '#C1C1C1',
    padding: '5px 9px'
  },
  input: {
    display: 'none'
  },
  replace: {
    marginLeft: 0,
    marginTop: 16
  },
  toolbar: {
    backgroundColor: '#ECEDF1',
    borderBottom: '1px solid #E0E1E6',
    borderTopLeftRadius: '5px',
    borderTopRightRadius: '5px',
    display: 'flex',
    justifyContent: 'flex-end',
    padding: '12px 8px'
  },
  tooltip: {
    position: 'relative',
    textAlign: 'right'
  },
  tooltipContent: {
    background: '#333131',
    borderRadius: '4px',
    color: '#FFFFFF',
    fontSize: '14px',
    lineHeight: '18px',
    marginTop: '-70px',
    padding: '7px 16px',
    position: 'absolute',
    right: '5px',
    textAlign: 'left',
    width: '300px'
  },
  tooltipArrow: {
    width: 0,
    height: 0,
    borderLeft: '10px solid transparent',
    borderRight: '10px solid transparent',
    borderTop: '10px solid #333131',
    marginBottom: -10,
    marginTop: '2px',
    position: 'absolute',
    right: '5px'
  }
});

Container.propTypes = {
  classes: PropTypes.object.isRequired,
  imageUrl: PropTypes.string.isRequired,
  hotspotColor: PropTypes.string.isRequired,
  multipleCorrect: PropTypes.bool.isRequired,
  onImageUpload: PropTypes.func.isRequired,
  onUpdateImageDimension: PropTypes.func.isRequired,
  onUpdateShapes: PropTypes.func.isRequired,
  outlineColor: PropTypes.string.isRequired,
  shapes: PropTypes.shape([]).isRequired
};

export default withStyles(styles)(Container);
