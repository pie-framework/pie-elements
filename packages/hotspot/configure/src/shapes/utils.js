import { CircleShape } from './circle';
import { PolygonShape } from './polygon';
import { RectangleShape } from './rectagle';

export const SUPPORTED_SHAPES = {
  CIRCLE: CircleShape.name,
  POLYGON: PolygonShape.name,
  RECTANGLE: RectangleShape.name,
  NONE: 'none',
};

export const SHAPE_GROUPS = {
  CIRCLES: 'circles',
  POLYGONS: 'polygons',
  RECTANGLES: 'rectangles',
};
