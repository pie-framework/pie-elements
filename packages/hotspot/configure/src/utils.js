import cloneDeep from 'lodash/cloneDeep';

const updateImageDimensions = (initialDim, nextDim, keepAspectRatio, resizeType) => {
  // if we want to keep image aspect ratio
  if (keepAspectRatio) {
    const imageAspectRatio = initialDim.width / initialDim.height;

    if (resizeType === 'height') {
      // if we want to change image height => we update the width accordingly
      return {
        width: nextDim.height * imageAspectRatio,
        height: nextDim.height
      }
    }

    // if we want to change image width => we update the height accordingly
    return {
      width: nextDim.width,
      height: nextDim.width / imageAspectRatio
    }
  }

  // if we don't want to keep aspect ratio, we just update both values
  return {
    width: nextDim.width,
    height: nextDim.height
  }
};

// referenceInitialValue = the initial value of the Stage
// referenceNextValue = the next value of the Stage
// currentValue = the value that has to be re-sized influenced by the changes that were made on the Stage
const getDelta = (referenceInitialValue, referenceNextValue, currentValue) => (referenceNextValue / referenceInitialValue) * currentValue;

const getUpdatedRectangle = (initialDim, nextDim, shape) => ({
  ...shape,
  width: getDelta(initialDim.width, nextDim.width, shape.width),
  height: getDelta(initialDim.height, nextDim.height, shape.height),
  x: getDelta(initialDim.width, nextDim.width, shape.x),
  y: getDelta(initialDim.height, nextDim.height, shape.y),
});

const getUpdatedPlygon = (initialDim, nextDim, shape) => ({
  ...shape,
  points: shape.points.map(point => ({
    x: getDelta(initialDim.width, nextDim.width, point.x),
    y: getDelta(initialDim.height, nextDim.height, point.y),
  }))
});

// initialDim = the initial dimensions: { width, height } of the Stage
// nextDim = the next dimensions: { width, height } of the Stage
// shapes = array of shapes that have to be re-sized and re-positioned
const getUpdatedShapes = (initialDim, nextDim, shapes) => {
  return shapes.map(shape => {
    if (shape.group === 'rectangles') {
      return getUpdatedRectangle(initialDim, nextDim, shape);
    }

    if (shape.group === 'polygons') {
      return getUpdatedPlygon(initialDim, nextDim, shape);
    }
  });
};

// converts shapes map to shapes array
// example:
// from: { rectangles: [r1], polygons: [p1, p2]}
// to: [{ ...r1, group: 'rectangles' }, { ...p1, group: 'polygons' }, { ...p2, group: 'polygons' }]
// if a shape has index defined, keep it, otherwise initialize it
// index is used for the UNDO function
const getAllShapes = (shapesMap) => {
  shapesMap = shapesMap || {};
  const shapesArray = [];
  const shapesKeys = Object.keys(shapesMap);

  return shapesKeys.length
    ? shapesKeys.reduce((acc, currentShapeKey) =>
        acc.concat(
          shapesMap[currentShapeKey]
            ? shapesMap[currentShapeKey].map((shape, index) => ({
              ...shape,
              group: currentShapeKey,
              index: shape.index || acc.length + index
            }))
            : []),
      shapesArray)
    : shapesArray;
};

// converts shapes array to shapes map
// is the reverse of getAllShapes function
// example:
// from: [{ ...r1, group: 'rectangles' }, { ...p1, group: 'polygons' }, { ...p2, group: 'polygons' }]
// to: { rectangles: [r1], polygons: [p1, p2]}
const groupShapes = (shapesArray) => {
  shapesArray = shapesArray || [];
  const shapesMap = {
    rectangles: [],
    polygons: []
  };

  if (shapesArray.length) {
    return shapesArray.reduce((acc, { group, ...shapeProps }) => {
      acc[group] = [...(acc[group] || []), shapeProps];
      return acc;
    }, shapesMap);
  }

  return cloneDeep(shapesMap);
};


export { updateImageDimensions, getUpdatedShapes, getAllShapes, groupShapes, getUpdatedRectangle, getUpdatedPlygon };
