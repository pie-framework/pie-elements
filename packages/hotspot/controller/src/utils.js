import isEqual from 'lodash/isEqual';

export const getCorrectResponse = (choices) => choices
  .filter(c => c.correct)
  .map(c => ({ id: c.id }))
  .sort();

export const isResponseCorrect = (question, session) => {
  const { shapes: { rectangles, polygons } } = question;
  const choices = [...rectangles, polygons];
  const correctResponse = getCorrectResponse(choices);

  if (session.answers.length) {
    return isEqual((session.answers || []).sort(), correctResponse);
  }
  return false;
};
