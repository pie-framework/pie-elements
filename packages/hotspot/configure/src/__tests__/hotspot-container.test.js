import { shallow } from 'enzyme';
import React from 'react';
import { Container } from '../hotspot-container';
import { getAllShapes, groupShapes } from '../utils';

const model = () => ({
  imageUrl: 'https://cdn.fluence.net/image/0240eb1455ce4c4bb6180232347b6aef_W',
  shapes: groupShapes(getAllShapes({
    rectangles: [
      {
        id: '0',
        height: 140,
        width: 130,
        x: 1,
        y: 1,
        correct: true
      }, {
        id: '1',
        height: 140,
        width: 130,
        x: 140,
        y: 1
      }, {
        id: '2',
        height: 140,
        width: 130,
        x: 280,
        y: 1
      }
    ],
    polygons: [
      {
        id: '3',
        points: [{ x: 1, y: 148 }, { x: 1, y: 288 }, { y: 288, x: 129 }, { y: 148, x: 129 }],
        correct: true
      }, {
        id: '4',
        points: [{ y: 151, x: 141 }, { y: 289, x: 141 }, { y: 289, x: 269 }, { x: 269, y: 151 }],
        correct: false
      }, {
        id: '5',
        points: [{ x: 279, y: 150 }, { x: 279, y: 289 }, { x: 407, y: 289 }, { x: 407, y: 150 }],
        correct: false
      }]
  })),
  dimensions: {
    height: 291, width: 410
  },
  hotspotColor: 'rgba(137, 183, 244, 0.65)',
  outlineColor: 'blue',
  multipleCorrect: true
});

describe('HotspotContainer', () => {
  let w,
    onImageUpload = jest.fn(),
    onUpdateImageDimension = jest.fn(),
    onUpdateShapes = jest.fn(),
    initialModel = model()
  ;

  beforeEach(() => {
    w = extras => {
      const props = {
        classes: {},
        dimensions: initialModel.dimensions,
        imageUrl: initialModel.imageUrl,
        multipleCorrect: initialModel.multipleCorrect,
        hotspotColor: initialModel.hotspotColor,
        outlineColor: initialModel.outlineColor,
        onUpdateImageDimension: onUpdateImageDimension,
        onUpdateShapes: onUpdateShapes,
        onImageUpload: onImageUpload,
        shapes: initialModel.shapes,
        ...extras
      };

      return shallow(<Container {...props} />);
    }
  });

  describe('render', () => {
    it('renders', () => {
      expect(w()).toMatchSnapshot();
    })
  });

  describe('logic', () => {
    let wrapper;
    let formattedShapes = [
      {
        id: '0',
        height: 140,
        width: 130,
        x: 1,
        y: 1,
        correct: true,
        group: 'rectangles',
        index: 0
      }, {
        id: '1',
        height: 140,
        width: 130,
        x: 140,
        y: 1,
        group: 'rectangles',
        index: 1
      }, {
        id: '2',
        height: 140,
        width: 130,
        x: 280,
        y: 1,
        group: 'rectangles',
        index: 2
      },
      {
        id: '3',
        points: [
          { x: 1, y: 148 },
          { x: 1, y: 288 },
          { y: 288, x: 129 },
          { y: 148, x: 129 }],
        correct: true,
        group: 'polygons',
        index: 3
      }, {
        id: '4',
        points: [
          { y: 151, x: 141 },
          { y: 289, x: 141 },
          { y: 289, x: 269 },
          { x: 269, y: 151 }],
        correct: false,
        group: 'polygons',
        index: 4
      }, {
        id: '5',
        points: [
          { x: 279, y: 150 },
          { x: 279, y: 289 },
          { x: 407, y: 289 },
          { x: 407, y: 150 }
        ],
        correct: false,
        group: 'polygons',
        index: 5
      }
    ];

    beforeEach(() => {
      wrapper = w();
    });

    it('state will contain formatted shapes', () => {
      expect(wrapper.instance().state.shapes).toEqual(expect.arrayContaining(formattedShapes));
    });

    it('onUpdateShapes with new added shape', () => {
      const newShape = {
        id: '7',
        height: 140,
        width: 130,
        x: 280,
        y: 1,
        group: 'rectangles'
      };

      wrapper.instance().onUpdateShapes([...formattedShapes, newShape]);

      expect(onUpdateShapes).toHaveBeenLastCalledWith({
        rectangles: [
          ...initialModel.shapes.rectangles,
          {
            id: '7',
            height: 140,
            width: 130,
            x: 280,
            y: 1,
          }
        ],
        polygons: initialModel.shapes.polygons
      })
    });

    it('onUpdateShapes with no shapes', () => {
      wrapper.instance().onUpdateShapes([]);

      expect(onUpdateShapes).toHaveBeenLastCalledWith({
        rectangles: [],
        polygons: []
      })
    });

    it('handleUndo if no new shape was added', () => {
      wrapper.instance().handleUndo();
      expect(onUpdateShapes).toHaveBeenLastCalledWith({
        ...initialModel.shapes,
        polygons: initialModel.shapes.polygons.slice(0, -1)
      });
    });

    it('handleUndo if new shape was added', () => {
      const newShape = {
        id: '7',
        height: 140,
        width: 130,
        x: 280,
        y: 1,
        index: 6
      };

      // first, add a new shape
      wrapper.instance().state.shapes = getAllShapes({
        ...initialModel.shapes,
        rectangles: [ ...initialModel.shapes.rectangles, newShape],
      });
      wrapper.instance().handleUndo();
      expect(onUpdateShapes).toBeCalledWith(initialModel.shapes);
    });

    it('handleClearAll', () => {
      wrapper.instance().handleClearAll();
      expect(onUpdateShapes).toHaveBeenLastCalledWith({
        rectangles: [],
        polygons: []
      });
    });
  });
});
