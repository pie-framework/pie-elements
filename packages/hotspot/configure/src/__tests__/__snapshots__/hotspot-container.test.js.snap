// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotspotContainer render renders 1`] = `
<div>
  <div
    className="undefined "
    onDragExit={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <div>
      <WithStyles(UploadControl)
        label="Replace Image"
        onInputClick={[Function]}
        onUploadImage={[Function]}
        setRef={[Function]}
      />
      <WithStyles(RawButton)
        disabled={false}
        label="Undo"
        onClick={[Function]}
      />
      <WithStyles(RawButton)
        disabled={false}
        label="Clear all"
        onClick={[Function]}
      />
    </div>
    <div>
      <WithStyles(Drawable)
        dimensions={
          Object {
            "height": 291,
            "width": 410,
          }
        }
        disableDrag={[Function]}
        enableDrag={[Function]}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        imageUrl="https://cdn.fluence.net/image/0240eb1455ce4c4bb6180232347b6aef_W"
        multipleCorrect={true}
        onUpdateImageDimension={[MockFunction]}
        onUpdateShapes={[Function]}
        outlineColor="blue"
        shapes={
          Array [
            Object {
              "correct": true,
              "group": "rectangles",
              "height": 140,
              "id": "0",
              "index": 0,
              "width": 130,
              "x": 1,
              "y": 1,
            },
            Object {
              "group": "rectangles",
              "height": 140,
              "id": "1",
              "index": 1,
              "width": 130,
              "x": 140,
              "y": 1,
            },
            Object {
              "group": "rectangles",
              "height": 140,
              "id": "2",
              "index": 2,
              "width": 130,
              "x": 280,
              "y": 1,
            },
            Object {
              "correct": true,
              "group": "polygons",
              "id": "3",
              "index": 3,
              "points": Array [
                Object {
                  "x": 1,
                  "y": 148,
                },
                Object {
                  "x": 1,
                  "y": 288,
                },
                Object {
                  "x": 129,
                  "y": 288,
                },
                Object {
                  "x": 129,
                  "y": 148,
                },
              ],
            },
            Object {
              "correct": false,
              "group": "polygons",
              "id": "4",
              "index": 4,
              "points": Array [
                Object {
                  "x": 141,
                  "y": 151,
                },
                Object {
                  "x": 141,
                  "y": 289,
                },
                Object {
                  "x": 269,
                  "y": 289,
                },
                Object {
                  "x": 269,
                  "y": 151,
                },
              ],
            },
            Object {
              "correct": false,
              "group": "polygons",
              "id": "5",
              "index": 5,
              "points": Array [
                Object {
                  "x": 279,
                  "y": 150,
                },
                Object {
                  "x": 279,
                  "y": 289,
                },
                Object {
                  "x": 407,
                  "y": 289,
                },
                Object {
                  "x": 407,
                  "y": 150,
                },
              ],
            },
          ]
        }
        strokeWidth={5}
      />
    </div>
    <div>
      <pure(HelpIcon)
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      />
    </div>
  </div>
</div>
`;
